### PR TITLE
Implement basic resolved dependency tree telemetry

### PR DESCRIPTION
**Pull request by natidea**
This change supercedes https://github.com/dotnet/project-system/pull/2879

This implements simplified telemetry to detect resolved dependency tree based on whether the corresponding snapshot has unresolved dependencies.  It includes a telemetry service that maintains some light state for each Target Framework to keep track whether all expected rules have been observed; this information is passed as a property of the telemetry event and can be used to determine if the 'resolved' event is fired too early (so sessions can be appropriately filtered).

**Customer scenario**

Dependency node population can be slow in some solutions. This telemetry can be used to get statistical information about how long it takes to populate the Dependencies node. The telemetry will be used as part of planned A/B testing to determine the best configuration of internal settings

**Bugs this fixes:** 

https://github.com/dotnet/project-system/issues/2703
[509162](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/509162)

**Workarounds, if any**

N/A

**Risk**

Low, this PR does not change dependency tree behavior, primary risk would be incorrect new telemetry.

**Performance impact**

Low, telemetry state only measures whether rules have been observed i.e. it does not maintain state for every node in the dependencies tree 

**Is this a regression from a previous update?** No

**Root cause analysis:**

N/A Planned, Partner requested

**How was the bug found?**

Partner requested

**Developer Notes**
This fires two possible telemetry events when TreeUpdate completes based on whether the corresponding snapshot has unresolved dependencies: 
```
vs/projectsystem/managed/treeupdated/unresolved
vs/projectsystem/managed/treeupdated/resolved
```
After resolved is fired, a flag is set that prevents anymore telemetry for that session (unless there is a reload). Hence sessions without reloads may have many "unresolved" events, and zero or one "resolved" event. Custom properties on the event are:
- Project: project guid
- ObservedAllRules: True | False

/cc @dotnet/project-system @abpiskunov @lifengl @adrianvmsft

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                _stopTelemetry = !hasUnresolvedDependency;
+                observedAllRules = ObservedAllRules();
+            }
+
+            if (_projectId == null)
+            {
+                InitializeProjectId();
+            }
+
+            _telemetryService.PostProperties($"{TelemetryEventName}/{(hasUnresolvedDependency ? UnresolvedLabel : ResolvedLabel)}",
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs]: Which culture does this get formatted with?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs [hunk 2]
```diff
@@ -32,6 +32,7 @@ internal class DependenciesProjectTreeProvider :
         IProjectTreeProvider,
         IDependenciesTreeServices
     {
+        private readonly IDependencyTreeTelemetryService _treeTelemetryService;
```
Comment (abpiskunov) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs]: >IDependencyTreeTelemetryService [](start = 25, length = 31)

Could you please follow same patter for imported components in this file? Add a readonly property below constructor ..  It is just easier to read if code follows same pattern in the file ..

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                SetProjectId(_telemetryService.HashValue(_project.FullPath));
+            }
+        }
+
+        // helper to support testing
+        internal void SetProjectId(string projectId)
+        {
+            _projectId = projectId;
+        }
+
```
Comment (abpiskunov) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs]: I would wrap it with just a property "ProjectId" and use its get and set

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        /// Fire telemetry when dependency tree completes an update
+        /// </summary>
+        /// <param name="hasUnresolvedDependency">indicates if the snapshot used for the update had any unresolved dependencies</param>
+        public void ObserveTreeUpdateCompleted(bool hasUnresolvedDependency)
+        {
+            bool observedAllRules;
+            lock (_stateUpdateLock)
+            {
+                if (_stopTelemetry) return;
+                _stopTelemetry = !hasUnresolvedDependency;
```
Comment (abpiskunov) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs]: >hasUnresolvedDependency [](start = 34, length = 23)

How you stop telemetry if project has unresolved dependency?

Comment (abpiskunov) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs]: Should we combine it with observedAllRules flag?

---
In reply to: [150085491](https://github.com/dotnet/project-system/pull/2944#discussion_r150085491) [](ancestors = 150085491)

**Pull request merged**